src/Providers/ElasticApmServiceProvider.php

```diff
@@ -200,4 +207,84 @@ protected function listenForQueries()
             app('query-log')->push($query);
         });
     }
+
+    protected function getCmdTransName($command)
+    {
+        return $command->command ? ('artisan ' . $command->command) : 'artisan help';
+    }
+
+    protected function getJobTransName($job)
+    {
+        return $job->job ? ($job->connectionName.'-job ' . json_decode($job->job->getRawBody())->displayName) : 'job';
+    }
+
+    protected function listenForCommands()
+    {
+        $this->app->events->listen(\Illuminate\Console\Events\CommandStarting::class, function ($commandStarting) {
+
+            app(Agent::class)->startTransaction(
+                $this->getCmdTransName($commandStarting)
+            );
+        });
+
+        $this->app->events->listen(\Illuminate\Console\Events\CommandFinished::class, function ($commandFinished) {
+
+            $agent = app(Agent::class);
+            $transaction = $agent->getTransaction(
+                $this->getCmdTransName($commandFinished)
+            );
+            $transaction->setResponse([
+                'finished' => true,
+                'status_code' => $commandFinished->exitCode,
+            ]);
+
+            $transaction->setUserContext([
+                'id' => getmyuid(),
+                'username' => get_current_user(),
+            ]);
+
+            $transaction->setMeta([
+                'result' => $commandFinished->exitCode,
+                'type' => 'COMMAND'
+            ]);
+
+            $transaction->setSpans(app('query-log')->toArray());
+
+            $transaction->stop();
+            $agent->send();
+        });
+    }
+
+    protected function listenForJobs()
+    {
+        $this->app->queue->before(function (JobProcessing $job) {
+            app(Agent::class)->startTransaction(
+                $this->getJobTransName($job)
+            );
+        });
+
+        $this->app->queue->after(function (JobProcessed $job) {
+            $agent = app(Agent::class);
+            $transaction = $agent->getTransaction(
+                $this->getJobTransName($job)
+            );
+            $transaction->setUserContext([
+                'id' => getmyuid(),
+                'username' => get_current_user(),
+            ]);
+            $transaction->setMeta([
+                'type' => 'JOB'
+            ]);
+            $transaction->setSpans(app('query-log')->toArray());
+
+            $transaction->stop();
+            $agent->send();
+        });
+
+        $this->app->queue->failing(function (JobFailed $job) {
+            $agent = app(Agent::class);
+            $agent->captureThrowable(new \Exception('failed job'));
```

Comment:
Looking at the docs, couldn't this be 

```
$agent->captureThrowable($job->exception);
```

https://laravel.com/docs/5.8/queues#dealing-with-failed-jobs

It also appears that the JobFailed parameter is considered and "event" rather than a "job". The JobFailed class has a job if you need access to it.

Reply:
You are right, thx, i will remove it, seems like i forgot to refactor some legacy code from a 5.1 app.

Reply:
This callback is exectued when a job fails after the maximum amount of retries. 
At this point, the original exception (that caused the job to fail and that we find in `$job->exception`) has already been sent to the APM via the normal exception handler (once per attempt, including this last exception).

I believe we want a new type of exception here, one that tells us that the job  will not be processed any further (i'm ignoring nightly failed-jobs recovery since it'S usually not always done and happens much later). Otherwise, we just duplicate the last job attempt's exception.

I have this problem on a prod system. We had no way to discern between sendSMS jobs that failed but recovered in a retry (normal) and jobs that failed 'for good' in their final attempt (usually meant  the provider is offline, network down ect) .

I considered introducing a new exceptionclass here, but  opted for KISS ;)

Reply:
That makes sense and I can agree with not creating a new exception class for now. The part that I'm still unsure of is the generic `failed job` message. I have not worked with jobs in APM, but it seems that differentiating the failures would be useful. Would it make sense to grab the message from the job being failed and append it here? Something like:

```
new \Exception('failed job: ' . $job->exception->getMessage())
```